### PR TITLE
[Enhancement] Allow setting micro batching parameters

### PR DESCRIPTION
**Pull request by jsemric**
<!--- Thanks for sending a pull request! Please make sure to read the contribution guidelines, then fill out the blanks below before requesting a code review. -->

## Description
<!--- Describe your changes in detail -->
<!--- Attach screenshots here if appropriate. -->
In this PR, I add an option to modify micro-batching parameters. ~~I achieve this by introducing a class `SavedBundleLoader` which in addition to loading bentoml service files can modify some of the service metadata while keeping it consistent with its configuration (which is used by other classes in this project). With this abstraction, it should be easy to extend the modification of other service parameters.~~

The logging message should display what are the current micro-batching parameters:
```
 ➜  bentoml serve-gunicorn --enable-microbatch --mb-max-latency 3333 --mb-max-batch-size 3333 IrisClassifier:20201202154246_C8DC0A                                                                                                                                   
[2020-12-02 16:09:46,684] INFO - Starting BentoML API server in production mode..                                                                                                                                                                                                  
[2020-12-02 16:09:46,855] INFO - get_gunicorn_num_of_workers: 3, calculated by cpu count                                                                                                                                                                                           
[2020-12-02 16:09:46,864] INFO - Running micro batch service on :5000                                                                                                                                                                                                              
       ...                                                                           
[2020-12-02 16:09:46,966] INFO - Micro batch enabled for API `predict` max-latency: 3333 max-batch-size 3333
```

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
<!--- If it is based on a conversation in slack channel, pls quote related messages here -->
The change allows modification of micro batching parameters after the image has been built as requested in issue #930 . To modify the parameters one can either set the following environmental variables `BENTOML_MB_MAX_BATCH_SIZE`, `BENTOML_MB_MAX_LATENCY` or run the CLI with the `--mb-max-batch-size`,  `--mb-max-latency` options.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
I run unit tests and API server integration tests. Further, I played with the CLI manually.

~~I also added unit tests for SavedBundleLoader and an integration test for the new CLI arguments.~~

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature and improvements (non-breaking change which adds/improves functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Code Refactoring (internal change which is not user facing)
- [ ] Documentation
- [ ] Test, CI, or build
<!--- [ ] Others: describe the type of change if it does not fit to categories listed -->

## Component(s) if applicable
- [x] BentoService (service definition, dependency management, API input/output adapters)
- [ ] Model Artifact (model serialization, multi-framework support)
- [x] Model Server (mico-batching, dockerisation, logging, OpenAPI, instruments)
- [ ] YataiService gRPC server (model registry, cloud deployment automation)
- [ ] YataiService web server (nodejs HTTP server and web UI)
- [ ] Internal (BentoML's own configuration, logging, utility, exception handling)
- [x] BentoML CLI

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- If you plan to update documentation or tests in follow-up, please note -->
- [x] My code follows the bentoml code style, both `./dev/format.sh` and
  `./dev/lint.sh` script have passed
  ([instructions](https://github.com/bentoml/BentoML/blob/master/DEVELOPMENT.md#style-check-and-auto-formatting-your-code)).
- [ ] My change reduces project test coverage and requires unit tests to be added
- [x] I have added unit tests covering my code change
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: bentoml/saved_bundle/loader.py [hunk 1]
```diff
@@ -94,17 +218,7 @@ def _resolve_remote_bundle_path(bundle_path):
...
-            return load_saved_bundle_config(local_bundle_path)
-
-    try:
-        return SavedBundleConfig.load(os.path.join(bundle_path, "bentoml.yml"))
-    except FileNotFoundError:
-        raise BentoMLException(
-            "BentoML can't locate config file 'bentoml.yml'"
-            " in saved bundle in path: {}".format(bundle_path)
-        )
+    return SavedBundleLoader(bundle_path).load_saved_bundle_config()
```
Comment (jsemric) [inline bentoml/saved_bundle/loader.py:221]: I kept those functions for backward compatibility.

**Comment by parano** [comment]
thanks for the PR @jsemric, it looks like there are conflicts with the master branch, you will need to merge/rebase with master and resolve the conflicts

File: bentoml/saved_bundle/loader.py [hunk 2]
```diff
@@ -32,6 +32,130 @@
...
+                    local_bundle_path, loader.api_config_override
+                )
+                return method(new_loader, *args)
+
+        return method(loader, *args)
+
+    return wrapper
+
+
+class SavedBundleLoader:
```
Comment (parano) [inline bentoml/saved_bundle/loader.py:49]: I really like the refactoring of this module here, it makes things a lot cleaner.

Although I feel adding `api_config_override` to the loader or the BentoService class is probably not the best way to implement this feature here.  It probably makes more sense to just make it an additional parameter of the API server class

Comment (parano) [reply bentoml/saved_bundle/loader.py:49]: I'd suggest make `mb_max_latency` and `mb_max_batch_size`  two optional parameters of MarshalService and GunicornMarshalServer, and pass them down from CLI args here: https://github.com/bentoml/BentoML/blob/master/bentoml/server/__init__.py

cc @bojiang thoughts on this?

Comment (jsemric) [reply bentoml/saved_bundle/loader.py:49]: Ah, ok. I thought we want to keep BentoService API metada compatible with what is in MasrshalService. I will do it then how you suggest as it is much simpler. I will also submit the refactoring of the bundle loader in a separate PR.

File: bentoml/marshal/marshal.py [hunk 3]
```diff
@@ -186,10 +190,24 @@ def add_batch_handler(self, api_name, max_latency, max_batch_size):
...
+                )
+                max_batch_size = (
+                    self.mb_max_batch_size
+                    or api_pb.mb_max_batch_size
+                    or self.DEFAULT_MAX_BATCH_SIZE
+                )
                 self.add_batch_handler(api_pb.name, max_latency, max_batch_size)
-                logger.info("Micro batch enabled for API `%s`", api_pb.name)
+                logger.info(
+                    "Micro batch enabled for API `%s` max-latency: %s"
```
Comment (parano) [inline bentoml/marshal/marshal.py:205]: 👍

File: bentoml/saved_bundle/loader.py (hunk 2 above)
Comment (parano) [reply bentoml/saved_bundle/loader.py]: Sounds great! Thanks @jsemric

**Pull request merged**